### ISSUE TITLE
Workshop / Marketplace for publishing arbitrary content

### ISSUE THREAD

**Issue by Frooxius** [Yellow-Dog-Man/Resonite-Issues#1473]
### Is your feature request related to a problem? Please describe.

Currently the only type of content that can easily be published and be searchable on Resonite are worlds. Publishing various items - from avatars, gadgets, tools to various resources like ProtoFlux nodes, materials, textures and so on is only possible via sharing of public folders.

This has a number of issues:
- Poor discoverability - you need to find a world with them or get them off another person
- They cannot be searched or filtered

### Describe the solution you'd like

Our plan is to generalize the submission system to allow publishing of arbitrary types of items, instead of just worlds. This would work similarly to the Worlds tab, but with extensions - better UI for managing and updating the submissions for content creators.

World submissions would then become simply a subset of the content on the marketplace, along with a number of other items. 

Various types of items could also be integrated into other tools and UI's, giving users quick and easy access to the repository of user creations:
- Quickly search & equip avatars from all published ones
- Material tool can search and equip published materials as part of its workflow and show a quick library of public materials
- Easily search & set other assets - textures, sound effects, fonts...
- Easily find Facets, custom video/audio players, color dialogs and other pieces of UI
- Easily find assets for building worlds

This will also have good synergy with additional features that will be implemented in the future:
- License system - allow publishing of paid items and selling them directly on the platform
-- E.g. if user proves they have license to particular avatar (e.g. by providing a key or FBX file to unlock the license), they get access to all the pre-setup versions of those avatars
- Molecule (https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1212) will allow for easy versioning and version tracking
- Nested ProtoFlux nodes (https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/564) - easily search for custom nodes for various functionality - sharing nodes for handling anything will become very easy and could be integrated into the node selection itself

### Describe alternatives you've considered

- Utilizing Steam's workshop, but this would make this inaccessible to users outside of Steam and would not be as well integrated.
- Continuing to use public folders, but this would be hard to support with search functionality and others

**Comment by jligeza** [comment]
Comming from #1564 , I hope this feature receives high development priority. There are lots of hidden gems uploaded by users, which are hard to find and explore.

Regarding features not mentioned by OP, the items in the workshop / marketplace could be discoverable by tags, where anybody should be able to add any tag to any item (even those items they don't own), with a help of autocomplete feature, to help at content discovery.

Imho, content discovery is the most missing feature in Resonite currently.

**Comment by shiftyscales** [comment]
@jligeza - our current priority will be either of performance or IK as dictated by the results of https://github.com/Yellow-Dog-Man/Resonite-Issues/discussions/2066.

There are many things that would ultimately be important to Resonite's success- content discoverability is one- but so are many other things.

We have a generally good idea of things that would benefit the platform as users of the platform ourselves, alongside what our community requests here on GitHub https://github.com/Yellow-Dog-Man/Resonite-Issues/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc.

As-needed too, we can also continue to poll users to determine what the broader community considers to be the most impactful things to implement on the platform.

As linked in this issue- there are other features that would go well with the workshop. There are also other items that would benefit it as well, e.g. fuzzy searching #536 which would make it easier to find items by name.

The inventory UI rework #1119 is also likely to have a lot of overlap with the needs of this issue for setting up a framework to better sort, filter, and view differing kinds of content on Resonite.

**Comment by jligeza** [comment]
Workshop isn't even on second page of upvoted suggestions at the moment. Is it prioritised so low? 

It would be a crucial feature especially for new users who could use it to find their first avatar, at day one in Resonite.

If Froox could implement at least some ugly MVP with basic content tagging, it would be a huge boost for community. We could start tagging items before final refined implementation. No need for anything fancy in the beginning, just tag items and search them.

**Comment by Frooxius** [comment]
If you want this to have higher priority, best thing you can do is vote for this and convince others to do the same. We'd like to prioritize this at some point, because it will help the platform overall, but it's a hard thing to balance with other stuff.

We're not going to do ugly MVP - we avoid these from the principle, because they cause more work and headaches than they really save. 

We already have system for tagging & searching - it's used for worlds. That part is already implemented and it will be extended for the items as well, so that doesn't really help here, because that system already exists.

The main part that needs to be done is extending the system to provide UI for arbitrary items and have a good submission mechanism. There's not really way to do this quick and dirty without actually implementing the Workshop.

**Comment by jligeza** [comment]
This is how I envision an MVP (this isn't really a workshop, but only tagging items for content discovery):

To add tags:
1. Open a folder in inventory that is public/shared.
2. Select an item.
3. Click new button at the top called "tags".
4. Dialog opens where a user can insert comma separated tags with autocomplete feature (try reuse tags).
5. Confirm changes; tags were added.

To search by tags:
1. Go to a new tab "items" (just like RedX, effectively replacing it).
2. Insert comma separated tags into a text input with an autocomplete feature.
3. Click search; receive paginated results in a grid.
4. (Optional) Add an option to search for items without tags, so volunteers can easily find and tag them.

The tags can be later reused for a complete workshop feature, no effort is wasted.

**Comment by Frooxius** [comment]
We don't plan on implementing it like that, because that doesn't match our goals or existing systems and designs in place.

It in fact would waste a lot of effort on our end and add a lot of unnecessary one.

**Comment by lynrayy** [comment]
That's awesome!
Sorry that didn't read the full post, too much letters for me x3

**Comment by Mhowser** [comment]
I would like to see some sort of a review or commenting section functionality too. Let's say a creator's item listed on the marketplace breaks for whatever reason, other players can leave a review on it notifying the original creator and they can update it.

**Comment by SunSerega** [comment]
I'm really afraid paid items would diminish the tinkering and just openness of the platform.
But I think a licensing system would be needed to adapt Resonite for a higher player count.

One way I see to minimize such issues is to affect the defaults people would use for licenses.
Most importantly, I think it should be convenient to choose open-source and dual licenses (for personal vs monetized use).
GitHub itself is a good example - when you create a repo, it gives you license suggestions.
But I'm specifically thinking towards the dual licenses being one of the default options for Resonite because that would work much better for avatars than pure open-source licenses.

Even if you, in the end, don't agree with what I think should be the default, please put lots of thought into how the license defaults you choose would affect the platform.

**Comment by Frooxius** [comment]
Yeah I think the defaults will do a lot, as they'll nudge people towards certain options and use-cases.

I don't think it'll diminish the tinkering and openness of the platform though - we'll still want to promote free items and encourage people to keep their items open for tinkering and remixing.

**Comment by dalekcoffee** [comment]
Agree with Lynnray 
It would be awesome to have the workshop/Marketplace but also having the ability to link to the item and maybe even facilitating the copying / purchase of the item in the physical space would be great for the experience and immersion.

The thought of shared spaces like malls, personal shop worlds, and even small popup shops at events sounds really cool.

And for those afraid of a pricetag, I use shop mostly as comparable reference, I don't necessarily mean things HAVE to be paid, could have a popup stand of free items at events as well I'm just picturing possibilities :3
